File: libs/csharp/common/src/Core/Nodes/SourceRoot.cs
using Crosslight.Core.Nodes;
using Crosslight.Core.Utilities;
using System.Runtime.InteropServices;
using Crosslight.Core;
using static Crosslight.Core.ILanguage;

namespace Crosslight.src.Core.Nodes
{
    /// <summary>
    /// Textual source file root.
    /// </summary>
    public struct SourceRoot : INodePayload
    {
        private struct Imported
        {
            public nint FileName;
        }

        /// <summary>
        /// Name of the file that contains this root in UTF-8 format. Can be empty or <see langword="null"/>.
        /// </summary>
        public string? FileName;

        /// <summary>
        /// Create a new managed instance of the source root payload from the native representation.
        /// </summary>
        /// <param name="pointer">The native pointer to a source root payload.</param>
        public SourceRoot(nint pointer)
        {
            Imported imported = Marshal.PtrToStructure<Imported>(pointer);

            FileName = imported.FileName == 0
                ? null
                : Marshal.PtrToStringUTF8(imported.FileName);
        }

        /// <inheritdoc/>
        public readonly NodeType Type => NodeType.SourceRoot;

        /// <inheritdoc/>
        public readonly nint ToPointer(AcquireDelegate? acquire)
        {
            acquire ??= Marshal.AllocCoTaskMem;
            nint pointer = acquire(Marshal.SizeOf<Imported>());

            if (pointer == 0)
            {
                return 0;
            }

            nint fileName = FileName == null
                ? 0
                : Utf8String.ToPointer(FileName, acquire);

            Imported imported = new()
            {
                FileName = fileName,
            };

            Marshal.StructureToPtr(imported, pointer, false);

            return pointer;
        }

        public readonly object? AcceptVisitor(Node node, INodePayloadVisitor visitor)
        {
            return visitor.VisitSourceRoot(node, this);
        }

        public override string ToString()
        {
            return $"{{ {nameof(SourceRoot)}-{FileName} }}";
        }
    }
}

File: libs/cpp/lang/src/builders/source_root.cpp
#include "lang/builders/source_root.hpp"
#include "lang/builders/allocator.hpp"
#include "lang/builders/builder.hpp"
#include "lang/utils.hpp"
#include "core/node.h"
#include "core/nodes/node_type.h"

namespace b = cl::lang::builders;

b::builder
b::source_root(const b::allocator& m, const char *filename) {
    return m.source_root(filename);
}

b::builder
b::allocator::source_root(const char *filename) const {
    auto* payload = static_cast<struct cl_node_source_root*>(acquire(sizeof(struct cl_node_source_root)));

    if (payload == nullptr) {
        return { *this, nullptr, nullptr };
    }

    payload->file_name = utils::string_duplicate(filename, acquire);

    return builder::from_payload(*this, payload, ::source_root);
}
